File: app/src/domain/foods.ts
/**
 * Seed food list.
 *
 * Portions in natural units — "1 katori", "1 roti" — never grams. Asking for grams is
 * the largest single source of entry friction and of the guessing noise that biases the
 * expenditure measurement downward (REQUIREMENTS §5.2, §10.3).
 *
 * ~70 items here; §5.2 puts the shipping target at 150–300. This is scaffolding: after
 * about two weeks a user's own history is their database.
 */

export interface Food {
  n: string;
  u: string;
  k: number;
}

export const FOODS: Food[] = [
  { n: 'Roti / chapati', u: '1 roti', k: 120 },
  { n: 'Paratha, plain', u: '1 paratha', k: 210 },
  { n: 'Naan', u: '1 naan', k: 260 },
  { n: 'Rice, cooked', u: '1 katori', k: 205 },
  { n: 'Jeera rice', u: '1 katori', k: 240 },
  { n: 'Veg pulao', u: '1 katori', k: 230 },
  { n: 'Chicken biryani', u: '1 plate', k: 490 },
  { n: 'Bread slice', u: '1 slice', k: 75 },
  { n: 'Poha', u: '1 plate', k: 250 },
  { n: 'Upma', u: '1 plate', k: 270 },
  { n: 'Idli', u: '1 idli', k: 58 },
  { n: 'Dosa, plain', u: '1 dosa', k: 165 },
  { n: 'Masala dosa', u: '1 dosa', k: 290 },
  { n: 'Uttapam', u: '1 uttapam', k: 230 },
  { n: 'Medu vada', u: '1 vada', k: 145 },
  { n: 'Oats, cooked', u: '1 bowl', k: 160 },
  { n: 'Muesli with milk', u: '1 bowl', k: 230 },
  { n: 'Cornflakes with milk', u: '1 bowl', k: 210 },
  { n: 'Dal tadka', u: '1 katori', k: 180 },
  { n: 'Dal makhani', u: '1 katori', k: 280 },
  { n: 'Rajma', u: '1 katori', k: 215 },
  { n: 'Chole', u: '1 katori', k: 240 },
  { n: 'Sambar', u: '1 katori', k: 140 },
  { n: 'Paneer butter masala', u: '1 katori', k: 330 },
  { n: 'Palak paneer', u: '1 katori', k: 275 },
  { n: 'Mixed veg sabzi', u: '1 katori', k: 150 },
  { n: 'Aloo sabzi', u: '1 katori', k: 190 },
  { n: 'Bhindi masala', u: '1 katori', k: 165 },
  { n: 'Chicken curry', u: '1 katori', k: 290 },
  { n: 'Egg curry', u: '1 katori', k: 240 },
  { n: 'Fish curry', u: '1 katori', k: 250 },
  { n: 'Mutton curry', u: '1 katori', k: 350 },
  { n: 'Egg, boiled', u: '1 egg', k: 78 },
  { n: 'Omelette, 2 egg', u: '1 omelette', k: 220 },
  { n: 'Chicken breast, grilled', u: '1 piece', k: 165 },
  { n: 'Paneer, raw', u: '50 g', k: 145 },
  { n: 'Tofu', u: '100 g', k: 145 },
  { n: 'Curd / dahi', u: '1 katori', k: 100 },
  { n: 'Greek yogurt', u: '1 cup', k: 130 },
  { n: 'Whey protein', u: '1 scoop', k: 120 },
  { n: 'Peanuts', u: '1 handful', k: 170 },
  { n: 'Almonds', u: '10 almonds', k: 70 },
  { n: 'Samosa', u: '1 samosa', k: 260 },
  { n: 'Pakora', u: '4 pieces', k: 220 },
  { n: 'Bhel puri', u: '1 plate', k: 210 },
  { n: 'Pav bhaji', u: '1 plate', k: 400 },
  { n: 'Vada pav', u: '1 piece', k: 290 },
  { n: 'Veg sandwich', u: '1 sandwich', k: 250 },
  { n: 'Maggi noodles', u: '1 pack', k: 310 },
  { n: 'Marie biscuits', u: '4 biscuits', k: 95 },
  { n: 'Potato chips', u: '1 small pack', k: 160 },
  { n: 'Dark chocolate', u: '2 squares', k: 110 },
  { n: 'Tea with milk & sugar', u: '1 cup', k: 90 },
  { n: 'Coffee with milk', u: '1 cup', k: 80 },
  { n: 'Black coffee', u: '1 cup', k: 5 },
  { n: 'Milk, full fat', u: '1 glass', k: 150 },
  { n: 'Buttermilk', u: '1 glass', k: 60 },
  { n: 'Sweet lassi', u: '1 glass', k: 260 },
  { n: 'Orange juice', u: '1 glass', k: 110 },
  { n: 'Cola', u: '1 can', k: 140 },
  { n: 'Beer', u: '1 bottle', k: 145 },
  { n: 'Banana', u: '1 banana', k: 105 },
  { n: 'Apple', u: '1 apple', k: 95 },
  { n: 'Orange', u: '1 orange', k: 62 },
  { n: 'Mango', u: '1 mango', k: 200 },
  { n: 'Papaya', u: '1 bowl', k: 60 },
  { n: 'Green salad', u: '1 bowl', k: 45 },
  { n: 'Ghee', u: '1 tsp', k: 45 },
  { n: 'Cooking oil', u: '1 tbsp', k: 120 },
  { n: 'Butter', u: '1 tsp', k: 36 },
  { n: 'Sugar', u: '1 tsp', k: 16 },
];

export type Meal = 'breakfast' | 'lunch' | 'snack' | 'dinner';

export const MEALS: { k: Meal; label: string }[] = [
  { k: 'breakfast', label: 'Breakfast' },
  { k: 'lunch', label: 'Lunch' },
  { k: 'snack', label: 'Snack' },
  { k: 'dinner', label: 'Dinner' },
];

export const mealFor = (minutes: number): Meal =>
  minutes < 11 * 60 ? 'breakfast' : minutes < 16 * 60 ? 'lunch' : minutes < 19 * 60 ? 'snack' : 'dinner';

export const mealLabel = (m: Meal): string => MEALS.find((x) => x.k === m)?.label ?? 'Snack';

/**
 * Browse order: recents in recency order, then everything else lightest-first.
 *
 * Never length-capped. A cap on top of a lightest-first sort silently buries every heavy
 * item — biryani, samosa, pav bhaji — where browsing cannot reach them.
 */
export function browseGroups(
  recents: string[],
  custom: Food[],
): { label: string; items: Food[] }[] {
  const all = [...custom, ...FOODS];
  const rec = recents.map((n) => all.find((x) => x.n === n)).filter((x): x is Food => !!x);
  const mine = custom.filter((x) => !recents.includes(x.n));
  const rest = FOODS.filter((x) => !recents.includes(x.n)).sort((a, b) => a.k - b.k);
  return [
    { label: 'Recent', items: rec },
    { label: 'Your foods', items: mine },
    { label: 'All foods · lightest first', items: rest },
  ].filter((g) => g.items.length > 0);
}

export function search(q: string, recents: string[], custom: Food[]): Food[] {
  const needle = q.trim().toLowerCase();
  const hit = [...custom, ...FOODS].filter((x) => x.n.toLowerCase().includes(needle));
  const inRecents = hit.filter((x) => recents.includes(x.n));
  const others = hit.filter((x) => !recents.includes(x.n)).sort((a, b) => a.k - b.k);
  return [...inRecents, ...others];
}

File: mobile/lib/domain/foods.dart
/// Seed food list.
///
/// Portions in natural units by default — "1 katori", "1 roti". Asking for grams on a
/// roti is entry friction, and friction is what stops people logging (§5.2).
///
/// **But grams where grams are how the food is actually handled.** "Chicken breast,
/// 1 piece" is not a natural unit, it is a vague one — a piece is anywhere from 80 g to
/// 250 g, and that spread is exactly the guessing noise the natural-units rule exists to
/// avoid (§10.3). For meat, nuts, cheese and anything bought by weight, 100 g is both
/// more precise *and* less effort, because the packet already says it. The rule is
/// "whichever unit the user does not have to estimate", which is natural units for
/// cooked dishes and grams for weighed ingredients.
///
/// ~70 items here; §5.2 puts the shipping target at 150–300. This is scaffolding: after
/// about two weeks a user's own history is their database.
library;

class Food {
  final String n;
  final String u;
  final int k;
  const Food(this.n, this.u, this.k);

  Map<String, dynamic> toJson() => {'n': n, 'u': u, 'k': k};
  static Food fromJson(Map<String, dynamic> j) =>
      Food(j['n'] as String, j['u'] as String, j['k'] as int);
}

const List<Food> seedFoods = [
  Food('Roti / chapati', '1 roti', 120),
  Food('Paratha, plain', '1 paratha', 210),
  Food('Naan', '1 naan', 260),
  Food('Rice, cooked', '1 katori', 205),
  Food('Jeera rice', '1 katori', 240),
  Food('Veg pulao', '1 katori', 230),
  Food('Chicken biryani', '1 plate', 490),
  Food('Bread slice', '1 slice', 75),
  Food('Poha', '1 plate', 250),
  Food('Upma', '1 plate', 270),
  Food('Idli', '1 idli', 58),
  Food('Dosa, plain', '1 dosa', 165),
  Food('Masala dosa', '1 dosa', 290),
  Food('Uttapam', '1 uttapam', 230),
  Food('Medu vada', '1 vada', 145),
  Food('Oats, cooked', '1 bowl', 160),
  Food('Muesli with milk', '1 bowl', 230),
  Food('Cornflakes with milk', '1 bowl', 210),
  Food('Dal tadka', '1 katori', 180),
  Food('Dal makhani', '1 katori', 280),
  Food('Rajma', '1 katori', 215),
  Food('Chole', '1 katori', 240),
  Food('Sambar', '1 katori', 140),
  Food('Paneer butter masala', '1 katori', 330),
  Food('Palak paneer', '1 katori', 275),
  Food('Mixed veg sabzi', '1 katori', 150),
  Food('Aloo sabzi', '1 katori', 190),
  Food('Bhindi masala', '1 katori', 165),
  Food('Chicken curry', '1 katori', 290),
  Food('Egg curry', '1 katori', 240),
  Food('Fish curry', '1 katori', 250),
  Food('Mutton curry', '1 katori', 350),
  Food('Egg, boiled', '1 egg', 78),
  Food('Omelette, 2 egg', '1 omelette', 220),
  Food('Chicken breast, grilled', '100 g', 165),
  Food('Paneer, raw', '50 g', 145),
  Food('Tofu', '100 g', 145),
  Food('Curd / dahi', '1 katori', 100),
  Food('Greek yogurt', '1 cup', 130),
  Food('Whey protein', '1 scoop', 120),
  Food('Peanuts', '100 g', 567),
  Food('Almonds', '10 almonds', 70),
  Food('Almonds, weighed', '100 g', 579),
  Food('Cashews', '100 g', 553),
  Food('Walnuts', '100 g', 654),
  Food('Cheese, cheddar', '100 g', 402),
  Food('Chicken, cooked', '100 g', 239),
  Food('Mutton, cooked', '100 g', 294),
  Food('Fish, cooked', '100 g', 206),
  Food('Prawns, cooked', '100 g', 99),
  Food('Soya chunks, dry', '100 g', 345),
  Food('Oats, dry', '100 g', 389),
  Food('Muesli', '100 g', 375),
  Food('Dates', '100 g', 277),
  Food('Boiled potato', '100 g', 87),
  Food('Sweet potato, boiled', '100 g', 76),
  Food('Sweet corn', '100 g', 96),
  Food('Samosa', '1 samosa', 260),
  Food('Pakora', '4 pieces', 220),
  Food('Bhel puri', '1 plate', 210),
  Food('Pav bhaji', '1 plate', 400),
  Food('Vada pav', '1 piece', 290),
  Food('Veg sandwich', '1 sandwich', 250),
  Food('Maggi noodles', '1 pack', 310),
  Food('Marie biscuits', '4 biscuits', 95),
  Food('Potato chips', '1 small pack', 160),
  Food('Dark chocolate', '2 squares', 110),
  Food('Tea with milk & sugar', '1 cup', 90),
  Food('Coffee with milk', '1 cup', 80),
  Food('Black coffee', '1 cup', 5),
  Food('Milk, full fat', '1 glass', 150),
  Food('Buttermilk', '1 glass', 60),
  Food('Sweet lassi', '1 glass', 260),
  Food('Orange juice', '1 glass', 110),
  Food('Cola', '1 can', 140),
  Food('Beer', '1 bottle', 145),
  Food('Banana', '1 banana', 105),
  Food('Apple', '1 apple', 95),
  Food('Orange', '1 orange', 62),
  Food('Mango', '1 mango', 200),
  Food('Papaya', '1 bowl', 60),
  Food('Green salad', '1 bowl', 45),
  Food('Ghee', '1 tsp', 45),
  Food('Cooking oil', '1 tbsp', 120),
  Food('Butter', '1 tsp', 36),
  Food('Sugar', '1 tsp', 16),

  // ---- breads & grains
  Food('Roti, multigrain', '1 roti', 130),
  Food('Bajra roti', '1 roti', 110),
  Food('Jowar roti', '1 roti', 105),
  Food('Missi roti', '1 roti', 175),
  Food('Aloo paratha', '1 paratha', 300),
  Food('Paneer paratha', '1 paratha', 330),
  Food('Butter naan', '1 naan', 320),
  Food('Tandoori roti', '1 roti', 140),
  Food('Kulcha', '1 kulcha', 230),
  Food('Puri', '1 puri', 100),
  Food('Bhatura', '1 bhatura', 300),
  Food('Brown bread slice', '1 slice', 70),
  Food('Pav', '1 pav', 130),
  Food('Rice, brown cooked', '1 katori', 195),
  Food('Curd rice', '1 katori', 230),
  Food('Lemon rice', '1 katori', 250),
  Food('Fried rice, veg', '1 plate', 340),
  Food('Khichdi', '1 katori', 200),
  Food('Daliya', '1 katori', 160),
  Food('Quinoa, cooked', '1 katori', 140),
  Food('Pasta, cooked', '1 katori', 200),
  Food('Noodles, cooked', '1 katori', 220),

  // ---- south indian
  Food('Rava dosa', '1 dosa', 240),
  Food('Rasam', '1 katori', 60),
  Food('Coconut chutney', '2 tbsp', 90),
  Food('Appam', '1 appam', 120),
  Food('Puttu', '1 serving', 200),

  // ---- dals & curries
  Food('Dal fry', '1 katori', 170),
  Food('Chana masala', '1 katori', 220),
  Food('Kadhi', '1 katori', 160),
  Food('Shahi paneer', '1 katori', 330),
  Food('Malai kofta', '1 katori', 360),
  Food('Aloo gobi', '1 katori', 150),
  Food('Aloo matar', '1 katori', 170),
  Food('Baingan bharta', '1 katori', 160),
  Food('Mixed veg curry', '1 katori', 165),
  Food('Butter chicken', '1 katori', 340),
  Food('Chicken tikka masala', '1 katori', 320),
  Food('Tandoori chicken', '2 pieces', 260),
  Food('Chicken tikka', '4 pieces', 220),
  Food('Fish fry', '1 piece', 200),
  Food('Prawn curry', '1 katori', 230),
  Food('Keema', '1 katori', 300),

  // ---- snacks & street
  Food('Idli sambar', '2 idli', 180),
  Food('Sabudana khichdi', '1 plate', 300),
  Food('Dhokla', '2 pieces', 120),
  Food('Khandvi', '3 pieces', 110),
  Food('Chole bhature', '1 plate', 650),
  Food('Momos, veg', '6 pieces', 250),
  Food('Momos, chicken', '6 pieces', 290),
  Food('Spring roll', '1 roll', 180),
  Food('Frankie / roll', '1 roll', 350),
  Food('Chicken shawarma', '1 roll', 400),
  Food('Burger, veg', '1 burger', 350),
  Food('Burger, chicken', '1 burger', 450),
  Food('Pizza slice', '1 slice', 270),
  Food('French fries', '1 small', 230),
  Food('Sev puri', '1 plate', 230),
  Food('Pani puri', '6 pieces', 180),
  Food('Dahi puri', '1 plate', 260),
  Food('Aloo tikki', '1 piece', 150),
  Food('Kachori', '1 piece', 190),
  Food('Poppadom', '1 papad', 45),
  Food('Murukku', '2 pieces', 110),
  Food('Bhujia / namkeen', '1 handful', 150),
  Food('Popcorn, plain', '1 bowl', 100),

  // ---- eggs & protein
  Food('Egg white, boiled', '1 egg', 17),
  Food('Egg bhurji', '2 eggs', 240),
  Food('Egg, fried', '1 egg', 90),
  Food('Paneer, grilled', '100 g', 265),
  Food('Tofu, grilled', '100 g', 160),
  Food('Soya chunks, cooked', '1 katori', 180),
  Food('Peanut butter', '1 tbsp', 95),
  Food('Sprouts salad', '1 katori', 110),

  // ---- fruit & veg
  Food('Grapes', '1 bowl', 90),
  Food('Watermelon', '1 bowl', 45),
  Food('Pomegranate', '1 bowl', 85),
  Food('Guava', '1 guava', 65),
  Food('Pear', '1 pear', 100),
  Food('Chikoo', '1 chikoo', 95),
  Food('Pineapple', '1 bowl', 80),
  Food('Custard apple', '1 fruit', 145),
  Food('Coconut water', '1 glass', 45),
  Food('Cucumber', '1 bowl', 20),
  Food('Carrot', '1 carrot', 25),
  Food('Tomato', '1 tomato', 20),
  Food('Sprouted moong', '1 katori', 100),
  Food('Boiled corn', '1 cob', 100),

  // ---- dairy & drinks
  Food('Milk, toned', '1 glass', 110),
  Food('Milk, skimmed', '1 glass', 85),
  Food('Curd, low fat', '1 katori', 70),
  Food('Paneer tikka', '4 pieces', 250),
  Food('Cheese slice', '1 slice', 70),
  Food('Masala chai', '1 cup', 100),
  Food('Green tea', '1 cup', 2),
  Food('Cold coffee', '1 glass', 220),
  Food('Filter coffee', '1 cup', 90),
  Food('Salted lassi', '1 glass', 110),
  Food('Mango shake', '1 glass', 280),
  Food('Banana shake', '1 glass', 250),
  Food('Fresh lime soda', '1 glass', 90),
  Food('Iced tea', '1 glass', 90),
  Food('Diet cola', '1 can', 2),
  Food('Whisky, 30 ml', '1 peg', 70),
  Food('Wine', '1 glass', 125),

  // ---- sweets
  Food('Gulab jamun', '1 piece', 150),
  Food('Rasgulla', '1 piece', 106),
  Food('Jalebi', '2 pieces', 200),
  Food('Ladoo', '1 piece', 185),
  Food('Barfi', '1 piece', 130),
  Food('Halwa', '1 katori', 280),
  Food('Kheer', '1 katori', 250),
  Food('Ice cream', '1 scoop', 140),
  Food('Chocolate bar', '1 bar', 250),
  Food('Biscuit, cream', '2 biscuits', 100),
  Food('Rusk', '2 pieces', 100),
  Food('Cake slice', '1 slice', 300),
  Food('Jaggery', '1 tsp', 20),
  Food('Honey', '1 tsp', 22),

  // ---- fats & extras
  Food('Mayonnaise', '1 tbsp', 90),
  Food('Tomato ketchup', '1 tbsp', 20),
  Food('Olive oil', '1 tbsp', 120),
  Food('Cream', '1 tbsp', 50),
  Food('Pickle', '1 tsp', 25),
];

enum Meal { breakfast, lunch, snack, dinner }

extension MealLabel on Meal {
  String get label => switch (this) {
        Meal.breakfast => 'Breakfast',
        Meal.lunch => 'Lunch',
        Meal.snack => 'Snack',
        Meal.dinner => 'Dinner',
      };
}

Meal mealFor(int minutes) => minutes < 11 * 60
    ? Meal.breakfast
    : minutes < 16 * 60
        ? Meal.lunch
        : minutes < 19 * 60
            ? Meal.snack
            : Meal.dinner;

class FoodGroup {
  final String label;
  final List<Food> items;
  const FoodGroup(this.label, this.items);
}

/// Browse order: recents in recency order, then everything else lightest-first.
///
/// Never length-capped. A cap on top of a lightest-first sort silently buries every
/// heavy item — biryani, samosa, pav bhaji — where browsing cannot reach them.
/// Browse groups, optionally led by what this user usually eats at this hour.
///
/// [habitual] is food names ordered by how often they appear in the current meal, from
/// `domain/insights.dart`. Recency alone puts last night's biryani at the top of the
/// breakfast list; what people want is what they normally eat *now*. Passing an empty
/// list gives the old recency-only behaviour, which is what a new user gets.
List<FoodGroup> browseGroups(
  List<String> recents,
  List<Food> custom, {
  List<String> habitual = const [],
  String habitLabel = 'Usually now',
}) {
  final all = [...custom, ...seedFoods];
  Food? find(String n) {
    final hit = all.where((f) => f.n == n);
    return hit.isEmpty ? null : hit.first;
  }

  final usual = <Food>[for (final n in habitual) ?find(n)];
  final taken = usual.map((f) => f.n).toSet();

  final rec = <Food>[
    for (final n in recents)
      if (!taken.contains(n)) ?find(n),
  ];
  taken.addAll(rec.map((f) => f.n));

  final mine = custom.where((f) => !taken.contains(f.n)).toList();
  final rest = seedFoods.where((f) => !taken.contains(f.n)).toList()
    ..sort((a, b) => a.k - b.k);

  return [
    FoodGroup(habitLabel, usual),
    FoodGroup('Recent', rec),
    FoodGroup('Your foods', mine),
    FoodGroup('All foods · lightest first', rest),
  ].where((g) => g.items.isNotEmpty).toList();
}

List<Food> searchFoods(String q, List<String> recents, List<Food> custom) {
  final needle = q.trim().toLowerCase();
  final hit =
      [...custom, ...seedFoods].where((f) => f.n.toLowerCase().contains(needle)).toList();
  final inRecents = hit.where((f) => recents.contains(f.n)).toList();
  final others = hit.where((f) => !recents.contains(f.n)).toList()
    ..sort((a, b) => a.k - b.k);
  return [...inRecents, ...others];
}
